features/FEATURE_BLE/targets/TARGET_NORDIC/TARGET_NRF5/source/nRF5XPalGattClient.cpp

```diff
@@ -0,0 +1,1616 @@
+         */
+        struct CustomAttReadByGroupTypeResponse : AttReadByGroupTypeResponse {
+            CustomAttReadByGroupTypeResponse(const services_array_t &s) :
+                AttReadByGroupTypeResponse(), services(s) {
+            }
+
+            virtual size_t size() const
+            {
+                return services.size();
+            }
+
+            virtual attribute_data_t operator[](size_t i) const
+            {
+                attribute_data_t result = {
+                    to_ble_handle_range(services[i].handle_range),
+                    make_const_ArrayView(
+                        reinterpret_cast<const uint8_t*>(&services[i].uuid.uuid),
+                        sizeof(uint16_t)
+                    )
+                };
+                return result;
+            }
+
+            const services_array_t &services;
+        };
+
+        terminate(CustomAttReadByGroupTypeResponse(services));
+    }
+
+    /**
+     * Handle discovery of services with a 128 UUID.
+     *
+     * Handle of the services discovered will be stored locally then the
+     * procedure sequentially initiate a read request of each of these attribute
+     * handle to acquire the value of the UUID of the service.
+     *
+     * The procedure ends once all the informations initially sent by the peer
+     * has been reconstructed and forwarded to the registered client handler.
+     */
+    void handle_128bit_services_discovered(const services_array_t &services)
+    {
+        response = new(std::nothrow) packed_discovery_response_t[services.size()];
+        if (!response) {
+            abort();
+            return;
+        }
+
+        count = services.size();
+        idx = 0;
+        for (size_t i = 0; i < count; ++i) {
+            uint8_t *it = &response[i][0];
+            u16_to_stream(it, services[i].handle_range.start_handle);
+            u16_to_stream(it, services[i].handle_range.end_handle);
+        }
+
+        read_service_uuid();
+    }
+
+    /**
+     * Initiate the read request of the next service attribute handle.
+     */
+    void read_service_uuid(void)
+    {
+        // note: use read multiple once different mtu size are supported ?
+        uint16_t attribute_handle = bytes_to_u16(&response[idx][0]);
+
+        uint32_t err = sd_ble_gattc_read(connection_handle, attribute_handle, 0);
+        if (err) {
+            abort();
+        }
+    }
+
+    /**
+     * Handle reception of a service (long) UUID.
+     */
+    void handle_uuid_read(const ble_gattc_evt_t &evt)
+    {
+        if (evt.gatt_status != BLE_GATT_STATUS_SUCCESS) {
+            terminate(AttErrorResponse(
+                AttributeOpcode::READ_REQUEST,
+                convert_sd_att_error_code(evt.gatt_status)
+            ));
+            return;
+        }
+
+        const ble_gattc_evt_read_rsp_t &rsp = evt.params.read_rsp;
+
+        uint16_t expected_handle = bytes_to_u16(&response[idx][0]);
```

Comment:
do all those numbers like 0, 16, 4 mean something? maybe they should be consts

Reply:
Offset 0 is explicit enough otherwise we can add static const variable for loong uuid length (16) and long uuid offset in read by group type response.